src/mmr/mmr_plonky2_verifier.rs

```diff
@@ -0,0 +1,210 @@
+  nr_merkle_proof_elms: usize,
+  nr_peaks: usize
+  // Returns circuit data, targets for leaf, targets for proof elements (hashes), targets for peaks
+) -> (CircuitData<GoldilocksField, PoseidonGoldilocksConfig, 2>, Target, Vec<(HashOutTarget, BoolTarget)>, Vec<HashOutTarget>) {
+  const D: usize = 2;
+  type C = PoseidonGoldilocksConfig;
+  type F = <C as GenericConfig<D>>::F;
+  // Verifying proof does the following:
+  // 1. Hashes its way through the (public input) merkle proof elements
+  // 2. Check result of (1) is amongst peaks
+  //     (for this, compare it to all peaks and check that the OR of these comparisons together true)
+  // 3. Hash peaks and compare to public input root
+
+  let mut proof_targets: Vec<(HashOutTarget, BoolTarget)> = Vec::new();
+  let mut peak_targets: Vec<HashOutTarget> = Vec::new();
+
+  let config = CircuitConfig::standard_recursion_config();
+  let mut builder: CircuitBuilder<plonky2::field::goldilocks_field::GoldilocksField, 2> = CircuitBuilder::<F, D>::new(config);
+  // The leaf to prove is in the MMR
+  let leaf_to_prove = builder.add_virtual_target();
+  let hashed_leaf = builder.hash_or_noop::<PoseidonHash>([leaf_to_prove].to_vec());
+  
+  // The first hashing outside of the loop, since it uses the leaf_to_prove
+  let mut next_hash: plonky2::hash::hash_types::HashOutTarget = hashed_leaf;
+
+  let mut proof_elm_index = 0;
+  while proof_elm_index < nr_merkle_proof_elms {
+    let merkle_proof_elm = builder.add_virtual_hash();
+    let elm_on_left = builder.add_virtual_bool_target_safe();
+    proof_targets.push((merkle_proof_elm, elm_on_left));
+    // Create the 2 options and then chose the correct one
+    // Option 1: sibling on the left
+    let option1 = builder.hash_or_noop::<PoseidonHash>([
+      merkle_proof_elm.elements.to_vec(),
+      next_hash.elements.to_vec()
+    ].concat());
+    // Option 2: sibling on the right
+    let option2 = builder.hash_or_noop::<PoseidonHash>([
+      next_hash.elements.to_vec(),
+      merkle_proof_elm.elements.to_vec()
+    ].concat());
+
+    // Pick the right next hash according to the bool that has been given with this element
+    next_hash = pick_hash(&mut builder, option1, option2, elm_on_left);
+    proof_elm_index += 1;
+  }
+
+  // Hash all peaks together
+  let mut peaks: Vec<HashOutTarget> = Vec::new();
+  let mut equals: Vec<BoolTarget> = Vec::new();
+  for _peaks in 0..nr_peaks {
+    let peak = builder.add_virtual_hash();
+    peaks.push(peak);
+    peak_targets.push(peak);
+    let equals_peak: BoolTarget = equal(&mut builder, peak, next_hash);
+    equals.push(equals_peak);
+  }
+
+  // Now check that the resulting "next_hash" appears in the given peaks
+  let hash_in_peaks = or_list(&mut builder, equals);
+  // check that its "true"
+  let one: plonky2::iop::target::Target = builder.one();
+  builder.connect(one, hash_in_peaks.target); 
+  // for some reason this below doesn't work
+  // builder.assert_bool(hash_in_peaks);
```

Comment:
What is the expected behavior of this?

Reply:
I would expect it to add the check that the bool target is true, but it doesn't do that. Couldn't find out what it's supposed to do (whether it's a bug or by design) and also tried asking in Discord but didn't get an answer.